### PR TITLE
[MoveosStd] Refactor ObjectRef, context:new_object and context::remove_object

### PR DESCRIPTION
**Pull request by jolestar**
## Summary

[MoveosStd] Refactor ObjectRef, context:new_object and context::remove_object

1. Make `ObjectRef` can not `copy`.
2. `context::new_object<T>` auto add Object to storage, and return ObjectRef<T>
3. `context::remove_object<T>` auto unpack the Object, and return `(id, owner, T)`
4. Refactor examples to use ObjectRef.

Continue #935 

Part of #901 

@geometryolife Some documents and example readme need to be updated.
@wubuku the code generator needs to be updated with new Object and ObjectRef API.

### CODE REVIEW

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move [hunk 1]
```diff
@@ -131,15 +121,26 @@ module moveos_std::context {
...
     /// Create a new Object with owner
-    public fun new_object_with_owner<T: key>(self: &mut Context, owner: address, value: T): Object<T> {
-        object::new(&mut self.tx_context, owner, value)
+    /// Add the Object to the global object storage and return the ObjectRef
+    public fun new_object_with_owner<T: key>(self: &mut Context, owner: address, value: T): ObjectRef<T> {
+        let object_id = fresh_object_id(self);
+        new_object_with_id(self, object_id, owner, value)
+    }
+
+    public(friend) fun new_object_with_id<T: key>(self: &mut Context, id: ObjectID, owner: address, value: T) : ObjectRef<T> {
```
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/context.move:139]: If new an object and return `ObjectRef `by default, need to additionally call the `object_ref::id` method to get the object_id. Looking at the example,  only need the object_id in many cases, which will increase the burden to the developer.

**Comment by WGB5445** [comment]
ObjectRef is a way to divide Object permissions. ObjectRef allows anyone to operate on the Object body. Should we abstract the permissions of sending and extending Objects, so that Objects are more flexible?

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move (hunk 1 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:139]: Previously, developers needed to call `object::id(&obj)` to get ObjectID, now, replaced by `object_ref::id(&obj_ref)`.

**Comment by jolestar** [comment]
> ObjectRef is a way to divide Object permissions. ObjectRef allows anyone to operate on the Object body. Should we abstract the permissions of sending and extending Objects, so that Objects are more flexible?

Do you have some ideas about this?

**Comment by jolestar** [comment]
> Here, we limit the sending of Objects of type to the module of T. We can modify Objects so that they can be sent anywhere.
> 
> https://github.com/rooch-network/rooch/blob/aa85fd70355e8b9ad77a2004f22301185fe74489/moveos/moveos-stdlib/moveos-stdlib/sources/object.move#L56-#L60
> 
> We can limit the sending permissions of Objects by adding some fields. Only users who have TransferRef can send this Object.
> 
> ```
> struct Object {
>     allow_ungated_transfer: bool,
> }
> ```
> 
> As a result, all Objects can be sent arbitrarily, and the creator can close the sending of some Objects. Here is the implementation of Aptos, which I think is very good.
> 
> https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/object.move#L400-L410

We can provide a `public_transfer` to `Object`, if needed. The `public_transfer` requires the `T` has key + store ability but does not require `private_generics(T)` like Sui.

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move [hunk 2]
```diff
@@ -105,21 +106,10 @@ module moveos_std::context {
...
-        let object_id = object_ref::into_id(object_ref);
-        storage_context::remove<T>(&mut self.storage_context, object_id)
-    }
-
-    #[private_generics(T)]
-    /// Add object to object store
-    public fun add_object<T: key>(self: &mut Context, obj: Object<T>) {
-        storage_context::add<T>(&mut self.storage_context, obj)
+    /// Remove object from object store, and unpack the Object
+    public fun remove_object<T: key>(self: &mut Context, object_id: ObjectID): (ObjectID, address, T) {
```
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/context.move:110]: After remove object, the original ObjectRef reference still exists. If call `object_ref::borrrow` or `object_ref::borrrow_mut`, the underlying raw_table will throw a NotFound error code. The Rust life cycle management mechanism fails in this dimension. 

 Should we consider adding track ObjectRef?

**Review by baichuan3 — APPROVED** [review]

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move (hunk 2 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:110]: Record a TODO about this

https://github.com/rooch-network/rooch/blob/e79f988fbac899e7c60fdc745fac091fd9210a9f/moveos/moveos-stdlib/moveos-stdlib/sources/object_ref.move#L22-L25

Currently, before using ObjectRef, the dev needs to call `object_ref::exist_object` to ensure the object exists.

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move [hunk 3]
```diff
@@ -105,21 +106,10 @@ module moveos_std::context {
...
+        let obj = storage_context::remove<T>(&mut self.storage_context, object_id);
+        object::unpack_internal(obj)
     }
 
     public fun contains_object(self: &Context, object_id: ObjectID): bool {
```
[the remark this answers is not in the document]
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move]: Done

**Pull request merged**